tests/phpunit/test-class-wp-widget-media.php

```diff
@@ -0,0 +1,354 @@
+			),
+			$instance
+		);
+		$this->assertSame( $result, $instance );
+
+		// Should return valid attachment url.
+		$expected = array(
+			'url' => 'https://example.org',
+		);
+		$result   = $widget->update( $expected, $instance );
+		$this->assertSame( $result, $expected );
+
+		// Should filter invalid attachment url.
+		$result = $widget->update(
+			array(
+				'url' => 'not_a_url',
+			),
+			$instance
+		);
+		$this->assertNotSame( $result, $instance );
+
+		// Should return valid attachment title.
+		$expected = array(
+			'title' => 'What a title',
+		);
+		$result   = $widget->update( $expected, $instance );
+		$this->assertSame( $result, $expected );
+
+		// Should filter invalid attachment title.
+		$result = $widget->update(
+			array(
+				'title' => '<h1>W00t!</h1>',
+			),
+			$instance
+		);
+		$this->assertNotSame( $result, $instance );
+
+		// Should filter invalid key.
+		$result = $widget->update(
+			array(
+				'imaginary_key' => 'value',
+			),
+			$instance
+		);
+		$this->assertSame( $result, $instance );
+
+		add_filter( 'sanitize_text_field', array( $this, '_return_wp_error' ) );
+		$result = $widget->update(
+			array(
+				'title' => 'Title',
+			),
+			$instance
+		);
+		remove_filter( 'sanitize_text_field', array( $this, '_return_wp_error' ) );
+		$this->assertSame( $result, $instance );
+	}
+
+	/**
+	 * Helper function for Test_WP_Widget_Media::test_update().
+	 *
+	 * @return \WP_Error
+	 */
+	function _return_wp_error() {
+		return new WP_Error( 'some-error', 'This is not valid!' );
+	}
+
+	/**
+	 * Test create_link_for method.
+	 *
+	 * @covers WP_Widget_Media::create_link_for()
+	 */
+	function test_create_link_for() {
+		if ( version_compare( PHP_VERSION, '5.3', '<' ) ) {
+			$this->markTestSkipped( 'ReflectionMethod::setAccessible is only available for PHP 5.3+' );
+			return;
+		}
+
+		$attachment_id = self::factory()->attachment->create_object( DIR_TESTDATA . '/images/canola.jpg', 0, array(
+			'post_mime_type' => 'image/jpeg',
+		) );
+
+		$wp_widget_media = new ReflectionClass( 'WP_Widget_Media' );
+		$create_link_for = $wp_widget_media->getMethod( 'create_link_for' );
+		$create_link_for->setAccessible( true );
+
+		$result = $create_link_for->invokeArgs( $this->get_mocked_class_instance(), array(
+			get_post( $attachment_id ),
+		) );
+		$this->assertSame( '<a href="#"></a>',$result );
+
+		wp_update_post( array(
+			'ID' => $attachment_id,
+			'post_title' => 'Attachment Title',
+		) );
+
+		$result = $create_link_for->invokeArgs( $this->get_mocked_class_instance(), array(
+			get_post( $attachment_id ),
+			'file',
+		) );
+		$this->assertSame( '<a href="' . esc_url( wp_get_attachment_url( $attachment_id ) ) . '">Attachment Title</a>',$result );
+
+		$result = $create_link_for->invokeArgs( $this->get_mocked_class_instance(), array(
+			get_post( $attachment_id ),
+			'post',
+		) );
+		$this->assertSame( '<a href="http://example.org/?attachment_id=4">Attachment Title</a>',$result );
```

Comment:
This line fails for me locally:

```
1) Test_WP_Widget_Media::test_create_link_for
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-<a href="http://example.org/?attachment_id=4">Attachment Title</a>
+<a href="http://tests.wordpress-develop.dev/?attachment_id=4">Attachment Title</a>
```

One thing is we can't assume the site url is `http://example.org/`.  It can be specified [`wp-tests-config.php`](https://github.com/WordPress/wordpress-develop/blob/master/wp-tests-config-sample.php) to be something else by changing the `WP_TESTS_DOMAIN` constant, which is what I have done on my environment.

Secondly, the `?attachment_id=4` part here makes me nervous because it assumes there has been a fixed number of posts inserted before this point.

So should it not rather be:

```php
$this->assertSame( '<a href="' . esc_url( get_permalink( $attachment_id ) ) . '">Attachment Title</a>', $result );
```